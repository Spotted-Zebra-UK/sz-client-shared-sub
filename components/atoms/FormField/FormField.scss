@import '../../../variables';

.FormField {
  &__Label__Text {
    font-size: $p2-font-size;
    line-height: $small-line-height;
    text-align: left;
  }

  &__Label__Text--HasError {
    color: $purple;
  }

  &__Error {
    color: $purple;
    font-size: $p2-font-size;
    min-height: 16px;
    line-height: $small-line-height;
    margin-top: 8px;
    text-align: left;
  }
}
