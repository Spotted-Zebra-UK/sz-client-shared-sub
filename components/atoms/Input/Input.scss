@import '../../../variables';

.Input {
  border: 0;
  border-bottom: 2px solid $black;
  font-size: $p1-font-size;
  height: 36px;
  line-height: $medium-line-height;
  width: 100%;

  &:disabled {
    background-color: transparent;
  }
}
