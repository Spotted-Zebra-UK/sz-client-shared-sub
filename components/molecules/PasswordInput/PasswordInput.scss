@import '../../../variables';

.PasswordInput {
  border-bottom: 2px solid $black;
  display: flex;

  input {
    border: 0;
    font-size: $p1-font-size;
    height: 36px;
    line-height: $medium-line-height;
    width: calc(100% - 22px);
    font-size: 14px;
  }

  &__Button {
    border: 0;
    background-color: transparent;
    cursor: pointer;

    svg {
      width: 22px;

      path {
        fill: rgba(0, 0, 0, 0.16);
      }
    }
  }
}
