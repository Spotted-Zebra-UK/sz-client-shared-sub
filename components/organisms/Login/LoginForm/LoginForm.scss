.LoginForm {
  .FormField--email {
    margin-bottom: 8px;
  }

  .SquareButton {
    display: block;
    margin: 0 auto;
  }
}
