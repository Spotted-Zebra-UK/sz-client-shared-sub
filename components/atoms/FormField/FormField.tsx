import './FormField.scss';
import React, { FunctionComponent } from 'react';

interface IFormField {
  error?: string;
  fieldName: string;
  label?: string;
}

const FormField: FunctionComponent<IFormField> = ({
  children,
  error,
  label,
  fieldName,
}) => {
  return (
    <div className={`FormField FormField--${fieldName}`}>
      <label className="FormField__Label" htmlFor={fieldName}>
        {label ? (
          <div
            className={`FormField__Label__Text${
              error ? ' FormField__Label__Text--HasError' : ''
            }`}
          >
            {label}
          </div>
        ) : null}
        {children}
        <div className="FormField__Error">{error}</div>
      </label>
    </div>
  );
};

export default FormField;
