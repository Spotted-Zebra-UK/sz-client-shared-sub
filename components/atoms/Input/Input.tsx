import './Input.scss';
import React, { FunctionComponent } from 'react';

interface IInput {
  className?: string;
  name: string;
  onChange: (value: string, name: string) => void;
  placeholder?: string;
  value: string;
  type?: 'text' | 'password' | 'date' | 'number';
  isDisabled?: boolean;
  id?: string;
}

const Input: FunctionComponent<IInput> = ({
  className,
  name,
  onChange,
  placeholder,
  value,
  type,
  isDisabled,
  id,
}) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value, name);
  };

  return (
    <input
      className={`Input${className ? ` ${className}` : ''}`}
      name={name}
      onChange={handleChange}
      onBlur={handleChange}
      value={value}
      type={type || 'text'}
      disabled={isDisabled || false}
      placeholder={placeholder}
      id={id}
    />
  );
};

export default Input;
