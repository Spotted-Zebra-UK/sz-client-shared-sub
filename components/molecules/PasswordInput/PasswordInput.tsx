import './PasswordInput.scss';
import React, { FC, useState } from 'react';
import { ReactComponent as EyeIcon } from '../../../icons/Eye.svg';
import { ReactComponent as EyeSolidIcon } from '../../../icons/EyeSolid.svg';
import Button from '../../atoms/Button/Button';

interface IPasswordInput {
  name: string;
  onChange: (value: string, name: string) => void;
  placeholder?: string;
  value: string;
}

const PasswordInput: FC<IPasswordInput> = ({
  value,
  onChange,
  placeholder,
  name,
}) => {
  const [isVisible, setIsVisible] = useState(false);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value, name);
  };

  const handleVisibilityChange = () => {
    setIsVisible(prevIsVisible => !prevIsVisible);
  };

  return (
    <div className="PasswordInput">
      <input
        onChange={handleChange}
        name={name}
        placeholder={placeholder}
        value={value}
        type={isVisible ? 'text' : 'password'}
      />
      <Button
        onClick={handleVisibilityChange}
        className="PasswordInput__Button"
      >
        {isVisible ? <EyeSolidIcon /> : <EyeIcon />}
      </Button>
    </div>
  );
};

export default PasswordInput;
