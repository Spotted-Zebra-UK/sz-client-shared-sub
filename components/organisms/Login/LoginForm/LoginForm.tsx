import './LoginForm.scss';
import capitalize from 'lodash/capitalize';
import isEmpty from 'lodash/isEmpty';
import React, { FunctionComponent, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Button } from '@spotted-zebra-uk/sz-ui-shared.ui.button';
import { TextInputField } from '@spotted-zebra-uk/sz-ui-shared.widgets.text-input-field';
import validate from '../../../../helpers/validate';

interface ILoginFormValues {
  email: string;
  password: string;
}

interface ILoginFormErrors {
  email: string[];
  password: string;
}

interface ILoginForm {
  email?: string;
  onSubmit: (email: string, password: string) => void;
}

const LoginForm: FunctionComponent<ILoginForm> = props => {
  const { t } = useTranslation();
  const [values, setValues] = useState<ILoginFormValues>({
    email: props.email || '',
    password: '',
  });
  const [errors, setErrors] = useState<ILoginFormErrors | undefined>();

  const loginValidate = () => {
    return validate(values, {
      email: {
        email: {
          // TODO: Fix localization [EN-1930]
          message: `^${t('common.notValidEmail')}`,
        },
      },
      password: {
        presence: {
          allowEmpty: false,
          message: `^${t('common.emptyPassword')}`,
        },
      },
    });
  };

  const handleChange = (value: string, name: string) => {
    setValues(prevValues => ({ ...prevValues, [name]: value }));
    setErrors(prevErrors =>
      prevErrors ? { ...prevErrors, [name]: [] } : prevErrors
    );
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const newErrors = loginValidate();
    if (newErrors) {
      setErrors(newErrors);
    } else {
      props.onSubmit(values.email, values.password);
    }
  };

  return (
    <form className="LoginForm" onSubmit={handleSubmit}>
      <div className="LoginForm__Fields">
        <TextInputField
          id="email"
          label={capitalize(t('common.email'))}
          value={values.email}
          placeholder={capitalize(t('common.email'))}
          onChange={handleChange}
          ariaLabel={t('common.email')}
          hasError={!isEmpty(errors?.email)}
          bottomText={errors?.email?.join(' ')}
          type="email"
        />
        <TextInputField
          id="password"
          label={capitalize(t('common.password'))}
          value={values.password}
          placeholder={capitalize(t('common.password'))}
          onChange={handleChange}
          hasError={!isEmpty(errors?.password)}
          bottomText={errors?.password}
          ariaLabel={capitalize(t('common.password'))}
          type="password"
        />
      </div>
      {/* TODO: Fix localization [EN-1930] */}
      <Button type="submit" className="LoginForm__SubmitButton">
        Sign in
      </Button>
    </form>
  );
};

export default LoginForm;
