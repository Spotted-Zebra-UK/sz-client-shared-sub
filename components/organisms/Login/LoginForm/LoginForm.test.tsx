import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import FormField from '../../../atoms/FormField/FormField';
import Input from '../../../atoms/Input/Input';
import PasswordInput from '../../../molecules/PasswordInput/PasswordInput';
import LoginForm from './LoginForm';

describe('LoginForm', () => {
  let onSubmit: (email: string) => void;

  beforeEach(() => {
    onSubmit = jest.fn();
  });

  // it('should render correctly by default', () => {
  //   const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

  //   expect(toJson(wrapper)).toMatchSnapshot();
  //   expect(wrapper.find(FormField).at(0).props().error).toBeUndefined();
  //   expect(wrapper.find(Input).props().value).toBe('');
  //   expect(wrapper.find(FormField).at(1).props().error).toBeUndefined();
  //   expect(wrapper.find(PasswordInput).props().value).toBe('');
  // });

  it('should change password input value on input change', () => {
    const passwordValue = 'test-passwordValue';
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

    wrapper.find(PasswordInput).props().onChange(passwordValue, 'password');
    expect(wrapper.find(PasswordInput).props().value).toBe(passwordValue);
  });

  it('should change email input value on input change', () => {
    const emailValue = 'test@email.com';
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

    wrapper.find(Input).props().onChange(emailValue, 'email');
    expect(wrapper.find(Input).props().value).toBe(emailValue);
  });

  it('should call provided onSubmit', () => {
    const passwordValue = 'test-passwordValue';
    const emailValue = 'test@email.com';
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

    wrapper.find(PasswordInput).props().onChange(passwordValue, 'password');
    expect(wrapper.find(PasswordInput).props().value).toBe(passwordValue);

    wrapper.find(Input).props().onChange(emailValue, 'email');
    expect(wrapper.find(Input).props().value).toBe(emailValue);

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(eventMock.preventDefault).toHaveBeenCalled();
    expect(onSubmit).toHaveBeenCalledWith(emailValue, passwordValue);
  });

  it('should show error and not call onSubmit if submit too short password', () => {
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(eventMock.preventDefault).toHaveBeenCalled();
    expect(wrapper.find(FormField).at(1).props().error).toBe(
      'Password is too short (minimum is 5 characters)'
    );
    expect(onSubmit).not.toHaveBeenCalled();
  });

  it('should show error and not call onSubmit if submit invalid email', () => {
    const emailValue = 'invalid-email';
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

    wrapper.find(Input).props().onChange(emailValue, 'email');

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(eventMock.preventDefault).toHaveBeenCalled();
    expect(wrapper.find(FormField).at(0).props().error).toBe(
      'Not a valid email'
    );
    expect(onSubmit).not.toHaveBeenCalled();
  });

  it('should hide password error if error is visible and password input changed', () => {
    const passwordValue = 'test-passwordValue';
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(wrapper.find(FormField).at(1).props().error).toBe(
      'Password is too short (minimum is 5 characters)'
    );

    wrapper.find(PasswordInput).props().onChange(passwordValue, 'password');
    expect(wrapper.find(FormField).at(1).props().error).toBe('');
  });

  it('should hide email error if error is visible and email input changed', () => {
    const validEmailValue = 'test@email.com';
    const invalidEmailValue = 'test-emailValue';
    const wrapper = shallow(<LoginForm onSubmit={onSubmit} />);

    wrapper.find(Input).props().onChange(invalidEmailValue, 'email');

    const eventMock = {
      preventDefault: jest.fn(),
    };

    wrapper.find('form').simulate('submit', eventMock);

    expect(wrapper.find(FormField).at(0).props().error).toBe(
      'Not a valid email'
    );

    wrapper.find(Input).props().onChange(validEmailValue, 'email');
    expect(wrapper.find(FormField).at(0).props().error).toBe('');
  });
});
