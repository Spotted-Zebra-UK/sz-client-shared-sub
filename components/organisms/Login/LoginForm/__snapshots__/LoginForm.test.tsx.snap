// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render correctly by default 1`] = `
<form
  className="LoginForm"
  onSubmit={[Function]}
>
  <div
    className="LoginForm__Fields"
  >
    <FormField
      fieldName="email"
      label="Email"
    >
      <Input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        value=""
      />
    </FormField>
    <FormField
      fieldName="password"
      label="Password"
    >
      <PasswordInput
        name="password"
        onChange={[Function]}
        placeholder="Password"
        value=""
      />
    </FormField>
  </div>
  <SquareButton
    type="submit"
  >
    Sign in
  </SquareButton>
</form>
`;
